### PR TITLE
C++: Easier debugging of dataflow node `toString` output

### PR DESCRIPTION
**Pull request by MathiasVP**
One of the first things I do when I debug dataflow node-related things is to modify the `toString` of `InstructionNode`s and `OperandNode`s. This PR simplifies that process by creating an abstract class to control whether the debug version (or the efficient) version should be used.

Then, to enable debug printing, one can simply import `DebugPrinting.qll` to modify the output.

By default `toString` already present on `main` will be used. So nothing should change from the user point-of-view. I've checked that the `toString` still compiles to the efficient RA when debugging isn't enabled:
```ql
Evaluated relational algebra for predicate DataFlowPrivate::Node0Impl.toStringImpl/0#dispred#9d03b963@295a56td with tuple counts:
    5160283   ~0%    {2} r1 = JOIN DataFlowPrivate::OperandNode0#class#93408876_10#join_rhs WITH `Node0ToString::Node0ToString.normalOperandToString/1#dispred#e901e619#cpe#23` ON FIRST 1 OUTPUT Lhs.1, Rhs.1
                
    5699527   ~0%    {2} r2 = JOIN DataFlowPrivate::InstructionNode0#class#f561e555_10#join_rhs WITH `Node0ToString::Node0ToString.normalInstructionToString/1#dispred#2fbb829d#cpe#23` ON FIRST 1 OUTPUT Lhs.1, Rhs.1
                
  10859810  ~71%    {2} r3 = r1 UNION r2
                    return r3
```

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -6,6 +6,7 @@ private import semmle.code.cpp.ir.internal.IRCppLanguage
 private import SsaInternals as Ssa
 private import DataFlowImplCommon as DataFlowImplCommon
 private import codeql.util.Unit
+import Node0ToString
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:9]: Is there a reason thi sis not a private import?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:9]: Slightly, yes. It's needed [here](https://github.com/github/codeql/pull/15088/files#diff-e9e4c8dbaa54c9f16e1533c09bb66ff3ec456e747f98de139672cfa0412a463cR489). Alternatively, we can private import `Node0ToString` here _and_ in `DataFlowUtil`. But since `DataFlowPrivate` isn't exported out of the dataflow library (unlike `DataFlowUtil`) I think it's fine. If you disagree I'm happy to change it, though

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll:9]: Alright. I'll do that!

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll]: Fixed in fcc3113bfc7b0a543e670f2866345d2356a2cf1a. I also noticed that `isDebugMode` computed `isNotDebugMode`. That has now been fixed in 401ab3b0358708838912824dcb1316baad02f038

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Node0ToString.qll [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * should be printed as part of the `toString` output of dataflow nodes.
+ *
+ * To enable debug printing import the `DebugPrinting.ql` file. By default,
+ * non-debug output will be used.
+ */
+class Node0ToString extends Unit {
+  abstract predicate isDebugMode();
+
+  private string normalInstructionToString(Instruction i) {
+    not this.isDebugMode() and
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Node0ToString.qll:23]: Why is this line necessary?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/Node0ToString.qll:23]: Not strictly necessary. There was just no reason to evaluate `normalInstructionToString` if we're in debug mode (since we'll be evaluating the debug version of `toString`).

**Comment by MathiasVP** [comment]
DCA looks clean 🎉

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll (hunk 1 above)
Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowPrivate.qll]: Thanks.

**Review by jketema — APPROVED** [review]
